### PR TITLE
auto-recover: orphaned worker branch for #24796

### LINKED ISSUE #24796: worker dispatch should stop no-remote branch-orphan loops
## Summary

AwardsApp issue awardsapp/awardsapp#6535 burned many worker attempts on repeated `worker_branch_orphan` / `orphan worktree detected (crash_type=no_work, 0 commits)` outcomes before manual intervention.

## Evidence

- Issue #6535 accumulated many `WORKER_BRANCH_ORPHAN` comments across multiple branch names.
- `git ls-remote --heads origin 'feature/auto-*gh6535'` returned no remote branches during interactive recovery, despite comments saying workers had pushed branches.
- A later automated scan reported `orphan worktree detected (crash_type=no_work, 0 commits)`.
- Relevant aidevops areas: `.agents/scripts/pulse-dispatch-worker-launch.sh:1481-1500` and `.agents/scripts/shared-claim-lifecycle.sh:231-355`.

## Why this matters

When branch-orphan telemetry says a branch was pushed but the remote branch does not exist, redispatching the same issue is unlikely to succeed. The framework should classify the condition as infrastructure/recovery failure and stop burning worker attempts.

## Acceptance criteria

- Before redispatch after `worker_branch_orphan`, verify whether the named remote branch actually exists.
- If no remote branch exists or the worktree has zero commits, hold dispatch with a clear root-cause comment and next action instead of trying another standard worker.
- Include the relevant branch name, remote-branch probe, and worktree commit count in the blocker comment.
- Add/update tests for the "orphan comment but no remote branch" case and the "zero commits" case.

## Verification

- Run the focused dispatch-dedup / pulse-dispatch worker tests.
- Simulate an issue with repeated `WORKER_BRANCH_ORPHAN` comments where `git ls-remote` returns no branch and verify dispatch is held immediately.

---
[aidevops.sh](https://aidevops.sh) v3.20.61 plugin for [OpenCode](https://opencode.ai) v1.17.7 with gpt-5.5 spent 39m and 167,471 tokens on this with the user in an interactive session.

### PR DESCRIPTION
**Pull request by marcusquinn**
## Summary

- Holds worker redispatch when `WORKER_BRANCH_ORPHAN` telemetry points to a branch that no longer exists remotely.
- Holds worker redispatch when the recovered orphan branch has zero commits to preserve worker capacity and force maintainer recovery.
- Refactors the branch-orphan guard into smaller helpers so the fix passes the function-complexity regression gate.

## Root cause and systemic fix

Pulse previously treated repeated branch-orphan telemetry as a branch-specific retry signal, but it did not validate whether the named remote branch still existed or whether the recovered branch contained commits. That allowed repeated standard workers to be launched even when there was no actionable branch state to recover.

The systemic fix is to probe the remote branch and commit count before redispatch. Unrecoverable evidence now posts a clear blocker with the branch name, probe command, remote existence result, commit count, root cause, and next action.

## Files changed

- `.agents/scripts/dispatch-dedup-helper.sh`
- `.agents/scripts/pulse-dispatch-worker-launch.sh`
- `.agents/scripts/tests/test-worker-branch-orphan-loop.sh`

## Testing

- `.agents/scripts/tests/test-worker-branch-orphan-loop.sh`
- `shellcheck .agents/scripts/dispatch-dedup-helper.sh .agents/scripts/pulse-dispatch-worker-launch.sh .agents/scripts/tests/test-worker-branch-orphan-loop.sh`
- `.agents/scripts/complexity-regression-helper.sh check --base origin/main --head HEAD --metric function-complexity`
- `.agents/scripts/complexity-regression-helper.sh check --base origin/main --head HEAD --metric bash32-compat`
- `.agents/scripts/linters-local.sh`

## Runtime Testing

Runtime-verified via the focused shell regression harness, which exercises the dispatch helper with stubbed GitHub and git commands for repeated orphan comments, missing remote branches, and zero-commit orphan branches without launching production workers.

Resolves #24796

---
[aidevops.sh](https://aidevops.sh) v3.20.64 plugin for [OpenCode](https://opencode.ai) v1.17.7 with gpt-5.5 spent 21m and 195,291 tokens on this with the user in an interactive session.

### CODE REVIEW

**Review by alex-solovyev — DISMISSED** [review]
Auto-approved by pulse runner @alex-solovyev — author @marcusquinn confirmed collaborator, pre-merge gates passed.

**Review by vladimirdulov — DISMISSED** [review]
Auto-approved by pulse runner @vladimirdulov — author @marcusquinn confirmed collaborator, pre-merge gates passed.

File: .agents/scripts/dispatch-dedup-helper.sh [hunk 1]
```diff
@@ -103,6 +103,101 @@ _ddh_resolve_pr_base_branch() {
...
+	local commit_count="$unknown_value"
+
+	if [[ -n "$branch_name" && "$branch_name" != "HEAD" ]]; then
+		local remote_rc=0
+		remote_probe="git ls-remote --exit-code origin refs/heads/${branch_name}"
+		if [[ -n "$worktree_path" && ( -d "$worktree_path/.git" || -f "$worktree_path/.git" ) ]]; then
+			git -C "$worktree_path" ls-remote --exit-code origin "refs/heads/${branch_name}" >/dev/null 2>&1 || remote_rc=$?
+		else
+			git ls-remote --exit-code origin "refs/heads/${branch_name}" >/dev/null 2>&1 || remote_rc=$?
+		fi
```
[the remark this answers is not in the document]
Comment (marcusquinn) [reply .agents/scripts/dispatch-dedup-helper.sh:147]: @gemini-code-assist Verified. This was a PR-owned correctness issue: .agents/scripts/dispatch-dedup-helper.sh:127 now runs ls-remote with GIT_TERMINAL_PROMPT=0 and keeps stderr visible while discarding stdout only. Verification: .agents/scripts/tests/test-worker-branch-orphan-loop.sh passed; shellcheck .agents/scripts/dispatch-dedup-helper.sh .agents/scripts/tests/test-worker-branch-orphan-loop.sh passed.

File: .agents/scripts/dispatch-dedup-helper.sh [hunk 2]
```diff
@@ -103,6 +103,101 @@ _ddh_resolve_pr_base_branch() {
...
+			remote_exists="$unknown_value"
+			;;
+		esac
+	fi
+
+	if [[ -n "$worktree_path" && ( -d "$worktree_path/.git" || -f "$worktree_path/.git" ) ]]; then
+		commit_count=$(git -C "$worktree_path" rev-list --count "origin/${base_branch}..HEAD" 2>/dev/null || true)
+		if ! [[ "$commit_count" =~ ^[0-9]+$ ]]; then
+			commit_count=$(git -C "$worktree_path" rev-list --count HEAD 2>/dev/null || true)
+		fi
```
[the remark this answers is not in the document]
Comment (marcusquinn) [reply .agents/scripts/dispatch-dedup-helper.sh]: @gemini-code-assist Verified. This was a PR-owned correctness issue: .agents/scripts/dispatch-dedup-helper.sh:145 now counts origin/${base_branch}..origin/${branch_name} with an origin/${branch_name} fallback, so reused worktrees reset to base no longer report HEAD as zero. The diagnostic now reports Branch commit count. Verification: .agents/scripts/tests/test-worker-branch-orphan-loop.sh passed; shellcheck .agents/scripts/dispatch-dedup-helper.sh .agents/scripts/tests/test-worker-branch-orphan-loop.sh passed.

**Review by vladimirdulov — DISMISSED** [review]
Auto-approved by pulse runner @vladimirdulov — author @marcusquinn confirmed collaborator, pre-merge gates passed.

**Review by alex-solovyev — DISMISSED** [review]
Auto-approved by pulse runner @alex-solovyev — author @marcusquinn confirmed collaborator, pre-merge gates passed.

**Comment by marcusquinn** [comment]
CLAIM_RELEASED reason=clean runner=marcusquinn ts=2026-06-14T22:28:31Z aidevops_version=3.20.61 opencode_version=1.17.7 exit=0 session_count=1

**Review by alex-solovyev — APPROVED** [review]
Auto-approved by pulse runner @alex-solovyev — author @marcusquinn confirmed collaborator, pre-merge gates passed.

**Review by vladimirdulov — APPROVED** [review]
Auto-approved by pulse runner @vladimirdulov — author @marcusquinn confirmed collaborator, pre-merge gates passed.

**Comment by marcusquinn** [comment]
## Completion Summary

- **What**: Added branch-orphan recovery probes that hold unrecoverable redispatch loops and refactored the guard below the function-complexity gate.
- **Issue**: #24796
- **Files changed**: `.agents/scripts/dispatch-dedup-helper.sh`, `.agents/scripts/pulse-dispatch-worker-launch.sh`, `.agents/scripts/tests/test-worker-branch-orphan-loop.sh`
- **Testing**: focused orphan-loop test, ShellCheck on changed shell files, function-complexity regression, Bash 3.2 regression, and local linter suite.
- **Key decisions**: fail-open on telemetry/API parse uncertainty, but fail-closed with a blocker when local probes prove the remote branch is missing or the orphan branch has zero commits.


---
[aidevops.sh](https://aidevops.sh) v3.20.64 plugin for [OpenCode](https://opencode.ai) v1.17.7 with gpt-5.5 spent 21m and 195,291 tokens on this with the user in an interactive session.
